### ISSUE TITLE
AudioKit crashes on device, but not simulator

### ISSUE THREAD

**Issue by magnus80a** [audiokit/AudioKit#643]
In the below code AudioKit.start() crashes on my iPhone SE with iOS 10.1.1. It works fine on the Simulator.

```
private func play(note: Int) {
  let pluckedString = AKPluckedString()
  AudioKit.output = pluckedString
  AudioKit.start() // <-- Crash here!
  let frequency = note.midiNoteToFrequency()
  pluckedString.trigger(frequency: frequency)
}
```
The console error log is

```
2016-12-04 10:51:45.765130 MyApp[1833:720319] [aurioc] 889: failed: -10851 (enable 2, outf< 2 ch,      0 Hz, Float32, non-inter> inf< 2 ch,      0 Hz, Float32, non-inter>)
2016-12-04 10:51:45.766519 MyApp[1833:720319] [aurioc] 889: failed: -10851 (enable 2, outf< 2 ch,  44100 Hz, Float32, non-inter> inf< 2 ch,      0 Hz, Float32, non-inter>)
2016-12-04 10:51:45.767008 MyApp[1833:720319] [aurioc] 889: failed: -10851 (enable 2, outf< 2 ch,  44100 Hz, Float32, non-inter> inf< 2 ch,      0 Hz, Float32, non-inter>)
2016-12-04 10:51:45.767982 MyApp[1833:720319] [central] 54:   ERROR:    [0x1b42d7c40] >avae> AVAudioEngineGraph.mm:2515: PerformCommand: error -10851
```
What have I missed? I can't find any documentation about any needed extra setup for devices compared to the simulator. The version of AudioKit is 3.5. XCode version is 8.1

**Comment by aure** [comment]
You haven't missed anything.  It should work.  Is it just the plucked string model or would an oscillator also crash?  I'll investigate on my devices as well.

**Comment by magnus80a** [comment]
I tried with an Oscillator and get the same error. Works in simulator.
```
private func play(note: Int) {
    let waveform = AKTable(.sawtooth)
    let oscillator = AKOscillator(waveform: waveform)
    AudioKit.output = oscillator
    AudioKit.start() // <-- Crashes here
    oscillator.frequency = note.midiNoteToFrequency()
    oscillator.play()
  }
```

**Comment by ryanschneider** [comment]
I'm getting very similar crash playing back an .aif file I recorded with an AVAudioFile from an AVAudioInputNode tap (I'm in the process of converting from raw AVFoundation -> AudioKit):

```swift
guard let file = AKAudioFile(forReading: audioFileURL) else { return }
print(file, file.standard)
let player = try AKAudioPlayer(file: file)
AudioKit.output = player
AudioKit.start()
player.play()
```

I get this error:

```
<AVAudioFormat 0x170285140:  1 ch,  44100 Hz, Int16> false
[aurioc] 889: failed: -10851 (enable 2, outf< 2 ch,  44100 Hz, Float32, non-inter> inf< 2 ch,      0 Hz, Float32, non-inter>)
2016-12-05 10:21:32.477861 TestApp[32451:18491098] [central] 54:   ERROR:    [0x16e517000] >avae> AVAudioEngineGraph.mm:2515: PerformCommand: error -10851
2016-12-05 10:21:32.482728 TestApp[32451:18491098] *** Terminating app due to uncaught exception 'com.apple.coreaudio.avfaudio', reason: 'error -10851'
```

The file plays back just fine if, for example, I send it over to mac and use a standard app to play it back.

Is it because the file isn't "standard"?  In my case, it's a mono file (I suspect AKOscillator and PluckedString are mono too).  

I tried both:

```swift
guard let file = try AKAudioFile(
                forReading: self.filename,
                commonFormat: .pcmFormatFloat32,
                interleaved: false
            ) else { return }
```
And running the file through `file.exportAsynchronously(name: "test", baseDir: .temp, exportFormat: .wav) {...}` to get a "standard" file, but no luck. 

I'll try downgrading to 3.4 to see if the error still occurs.

**Comment by ryanschneider** [comment]
Also, FWIW, error `-10851 ` is `kAudioUnitErr_InvalidPropertyValue` according to osstatus.com:

https://www.osstatus.com/search/results?platform=all&framework=all&search=-10851

**Comment by ryanschneider** [comment]
Ok, I replaced the ViewController in the `HelloWorld` example with code that crashes for me in my app, and it played back just fine:

```swift
class ViewController: UIViewController {

    var player: AKAudioPlayer!

    @IBAction func toggleSound(_ sender: UIButton) {
        self.test()
    }

    func test() {
        do {
            let file = try AKAudioFile(readFileName: "rawMicrophone.aif", baseDir: .resources)
            self.player = try! AKAudioPlayer(file: file)

            AudioKit.output = self.player

            self.player.completionHandler = {
                print("Done!")
            }

            AudioKit.start()
            self.player.play()
        }
        catch let exception {
            print(exception)
            fatalError("TODO: Remove this and handle errors correctly!")
        }
    }
}
```

My next step will be to convert HelloWorld over to using CocoaPods to see if I now get the crash.

**Comment by ryanschneider** [comment]
Hmm, ok, well, I'm definitely at more of a loss now, as using the above view controller in a simple app where AudioKit is integrated via CocoaPods also works fine.

I'll try pulling in some more of my AVFoundation code to see if that's what is causing my issue.  If I find something I'll probably create a new issue and refer to this one rather than hijack it further if it appears to be something different than Magnus' issue.

**Comment by ryanschneider** [comment]
Ok, I found my issue, hopefully this helps Magnus as well.

When I was recording, I was setting the category to the Record category, basically doing something equivalent to:

```swift
try! AVAudioSession.sharedInstance().setCategory(AVAudioSessionCategoryRecord)
```

(But not using a `try!`).  Now when I go to play I get the -10851 error since the current category isn't compatible with playback.  So, it'd be great if AudioKit detected this by checking the category before starting playback, and either threw a Swift style error or other better error handling than an NSException, but at least in my case it's now really AudioKit's fault. :)

**Comment by aure** [comment]
Sorry I didn't notice this earlier, but clearly you were trying to start AudioKit too many times.

**Comment by magnus80a** [comment]
I'm using Cocoapods to install the framework.
When I get the chance I'll take a look at how I set the category on the audio session.
@aure  I get the crash on my first AudioKit.start() not my second so I don't agree that's necessarily the cause.

**Comment by magnus80a** [comment]
@ryanschneider It seems my problem is definitely related to your problem. By making sure that the aduio session category isn't AVAudioSessionCategoryRecord on playback; my app doesn't crash anymore. A big thank you! :)

**Comment by ryanschneider** [comment]
Glad it helped. :). -10851 seems to be Core Audio's catch all "something is
wrong" error code, I've hit it in several different scenarios today trying
to convert my code over to AudioKit.

**Comment by aure** [comment]
@ryanschneider Thanks for helping out.  @magnus80a Sorry for closing the issue early, luckily Ryan was here to save the day.

**Comment by LeeMatthewHiggins** [comment]
It seems my problem (the original ticket issue) was related to not holding onto the microphone object in my code. It must have worked before due to leaks. Strange that it works on the sim! I guess sim code must still have leaks, thats why it still worked. I wonder if the mic should be a singleton and be kept alive internally in audiokit?
